production.Dockerfile

```diff
@@ -1,69 +1,119 @@
+    "nodejs~=14" \
+    "npm~=7" \
+    && npm install -g yarn@1
 
+# Install SAML dependencies
+#
+# Notes:
+#
+# - please add in this section runtime dependences only.
+#   If you temporary need a package to build a Python or npm
+#   dependency take a look at the sections below.
+#
+# - we would like to include those dependencies + 'python3-saml'
+#   directly in the requirements.txt file but due to our CI/CD
+#   setup this is currently not possible. More context at:
+#   https://github.com/PostHog/posthog/pull/5870
+#   https://github.com/PostHog/posthog/pull/6575#discussion_r733457836
+#   https://github.com/PostHog/posthog/pull/6607
+#
+RUN apk --update --no-cache add \
+    "libxml2-dev~=2.9" \
+    "xmlsec~=1.2" \
+    "xmlsec-dev~=1.2" \
+    && \
+    pip install python3-saml==1.12.0 --compile --no-cache-dir
 
-# uninstall unneeded Python dependencies
-RUN pip uninstall ipython-genutils pip wheel -y
+# Compile and install Python dependencies.
+#
+# Notes:
+#
+# - we explicitly COPY the files so that we don't need to rebuild
+#   the container every time a dependency changes
+#
+# - we need few additional OS packages for this. Let's install
+#   and then uninstall them when the compilation is completed.
+COPY requirements.txt ./
+RUN apk --update --no-cache --virtual .build-deps add \
+    "cargo~=1.52" \
+    "git~=2" \
+    "libffi-dev~=3.3" \
+    "postgresql-dev~=13" \
+    && \
+    pip install -r requirements.txt --compile --no-cache-dir \
+    && \
+    apk del .build-deps
 
-# remove build dependencies not needed at runtime
-RUN  apt-get purge -y git curl build-essential && apt-get autoremove -y
+# Compile and install Yarn dependencies.
+#
+# Notes:
+#
+# - we explicitly COPY the files so that we don't need to rebuild
+#   the container every time a dependency changes
+#
+# - we need few additional OS packages for this. Let's install
+#   and then uninstall them when the compilation is completed.
+COPY package.json yarn.lock ./
+COPY plugins/package.json plugins/yarn.lock ./plugins/
+RUN apk --update --no-cache --virtual .build-deps add \
+    "gcc~=10.3" \
+    && \
+    yarn config set network-timeout 300000 && \
+    yarn install --frozen-lockfile && \
+    yarn install --frozen-lockfile --cwd plugins && \
+    yarn cache clean \
+    && \
+    apk del .build-deps
 
-# install JS (yarn) dependencies
-COPY package.json /code/.
-COPY yarn.lock /code/.
-RUN yarn --frozen-lockfile
+# Copy everything else
+COPY . .
 
-# steps below will change on almost every build (steps above will be cached most of the time)
-# load entire codebase & build frontend
-COPY . /code/
-RUN yarn build \
-    && yarn --cwd plugins --frozen-lockfile --ignore-optional \
-    && yarn cache clean \
-    && rm -rf node_modules
+# Build the frontend
+#
+# Note: we run the build as a separate actions to increase
+# the cache hit ratio of the layers above.
+RUN yarn build && \
+    yarn cache clean && \
+    rm -rf ./node_modules ./plugins/node_modules
```

Comment:
I may be missing something, but I don't think it's correct to delete `./plugins/node_modules`, as there's no `yarn build --cwd plugins` here?

Reply:
Plus we don't actually want to delete `./plugins/node_modules` at all as they are runtime deps for the plugin server?

We [install deps in `./bin/plugin-server` ](https://github.com/PostHog/posthog/blob/plugin-server-dates/bin/plugin-server#L43:L43)which is probably why it's not an issue but we probably shouldn't need to